test/inductor/test_cache.py

```diff
@@ -0,0 +1,150 @@
+
+            # Important to reset so in-memory guards don't prevent compilation
+            torch._dynamo.reset()
+
+            cached_result = foo(input, weight, bias)
+            cached_compilation_metrics = torch._dynamo.utils.compilation_metrics
+
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
+
+        def test_cache_module_output(self):
+            activate_unique_cache()
+
+            model = torch.compile(Foo())
+
+            input = torch.rand(15, 15)
+
+            base_result = model(input)
+            base_compilation_metrics = torch._dynamo.utils.compilation_metrics
+
+            # The compilation metrics of the first compilation should show codegen
+            self.assertIn("Scheduler.codegen", base_compilation_metrics)
+
+            # Important to reset so in-memory guards don't prevent compilation
+            torch._dynamo.reset()
+
+            cached_result = model(input)
+            cached_compilation_metrics = torch._dynamo.utils.compilation_metrics
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
+
+
+if HAS_CUDA and not TEST_WITH_ASAN:
+
+    class TestCodeGenCacheCuda(CacheTestCase):
+        def test_cache_cudagraphs(self):
+            activate_unique_cache()
+
+            model = Foo()
+
+            cuda_model1 = model.cuda()
+            opt_model1 = torch.compile(cuda_model1, mode="reduce-overhead")
+
+            input = torch.rand(15, 15).cuda()
+
+            base_result = opt_model1(input).clone().detach()
+            base_compilation_metrics = torch._dynamo.utils.compilation_metrics
+
+            # The compilation metrics of the first compilation should show codegen
+            self.assertIn("Scheduler.codegen", base_compilation_metrics)
+
+            # Important to reset so in-memory guards don't prevent compilation
+            torch._dynamo.reset()
+
+            cuda_model2 = model.cuda()
+            opt_model2 = torch.compile(cuda_model2, mode="reduce-overhead")
+            cached_result = opt_model2(input)
+            cached_compilation_metrics = torch._dynamo.utils.compilation_metrics
+            # The cached_result coming from a cached kernel should be the same as the original one
+            # The compilation metrics of the cached compilation should show no codegen
+            self.assertEqual(base_result, cached_result)
+            self.assertNotIn("Scheduler.codegen", cached_compilation_metrics)
```

Comment:
One of the more important classes of tests here will be ensuring that hash keys don't falsely match when they shouldn't. It will be good to think of a good strategy for this.